File: Pragma_Examples/OpenMP/C/5_device_routines/2_device_routine_wglobaldata/1_device_routine_wglobaldata/device_routine.c
// Copyright (c) 2024 AMD HPC Application Performance Team
// Author: Bob Robey, Bob.Robey@amd.com
// MIT License

#include <stdlib.h>
#include <stdio.h>

void compute(int cindex, double *x);

int main(int argc, char *argv[]){
   int N=1000;
   double *x = (double *)malloc(N*sizeof(double));
#pragma omp target enter data map (alloc:x[0:N])
#pragma omp target teams distribute parallel for
   for (int k = 0; k < N; k++){
      x[k] = 0.0;
   }

#pragma omp target teams distribute parallel for
   for (int k = 0; k < N; k++){
      int cindex = k%10;
      compute(cindex, &x[k]);
   }

   double sum = 0.0;
#pragma omp target teams distribute parallel for reduction(+:sum)
   for (int k = 0; k < N; k++){
      sum += x[k];
   }

   printf("Result: sum of x is %lf\n",sum);

#pragma omp target exit data map (release:x[0:N])
   free(x);
}
      

File: Pragma_Examples/OpenMP/Fortran/5_device_routines/device_routine_with_interface/2_device_routine_usm/device_routine.f90
! This example was created by Johanna Potyka
! Copyright (c) 2025 AMD HPC Application Performance Team
! MIT License

      program device_routine
      !----------------------------
      ! description: this program demonstrates
      !              how to call a device subroutine
         use omp_lib

         implicit none
         !$omp requires unified_shared_memory
         !---variables
         integer,parameter :: N=1000
         !N                   number of values in x array
         integer,parameter :: rk=kind(1.0d0)
         !rk                  kind of real
         integer :: k, err_stat
         !k         index
         !err_stat  status variable
         real(kind=rk),dimension(:),ALLOCATABLE :: x
         !x                                        array
         real(kind=rk) :: sum
         !sum             used to sum up x
         interface
                 subroutine compute(x)
                   integer,parameter :: rk=8
                   real(kind=rk), intent(inout) :: x
                 end subroutine compute
         end interface

         !---allocation
         allocate(x(1:N),STAT=err_stat)
         if(err_stat /= 0) then
             write(*,*) "error while allocating"
             STOP
         end if

         !---initialisation
         x = -1.0_rk
         !--- call a device subroutine in kernel
         !$omp target teams distribute parallel do
         do k=1,N
            call compute(x(k))
            !x(k) = 1.0_rk
         end do

         !--- initialize sum
        sum = 0.0_rk;

        !--- sum up x to sum on device with reduction
        !$omp target teams distribute parallel do reduction(+:sum)
        do k=1,N
           sum = sum + x(k)
        end do

        !--- print result
        Write(*,'(A,F0.12)') "Result: sum of x is ",sum

        deallocate(x)
      end program device_routine
